### PR TITLE
V22F-620 cover filedatastorage code

### PR DESCRIPTION
**Pull request by Sindarin27**
Cover all code of the FileDataStorage files.
Additionally, update the DecisionTuple class slightly to ensure Lights and Gear are usable, though not implemented.

### CODE REVIEW

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp [hunk 1]
```diff
@@ -35,31 +81,327 @@ inline std::string GetTimeAsString(time_t p_date)
...
+TEST(FileDataStorageTests, DecisionsDisabledNoDecisions)
+{
+    TestDataStorageSaveDecisions(false, false, false, false, false, false);
+}
+
+/// @brief Test whether the FileDataStorage stores properly when decisions should not be saved but all decisions are made
+TEST(FileDataStorageTests, DecisionsDisabledAllDecisions)
+{
+    TestDataStorageSaveDecisions(false, true, true, true, true, true);
+}
```
Comment (KungFuDonkey) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp:326]: Can be converted into testcases

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp [hunk 2]
```diff
@@ -35,31 +81,327 @@ inline std::string GetTimeAsString(time_t p_date)
...
+
+/// @brief Test whether the FileDataStorage stores properly for every combination of decision made
+TEST(FileDataStorageTests, DecisionsCombinations)
+{
+    bool alwaysTrue[1]{true};
+    bool booleans[2]{true, false};
+    PairWiseTest(TestDataStorageSaveDecisions, alwaysTrue, 1, booleans, 2, booleans, 2, booleans, 2, booleans, 2, booleans, 2);
+}
```
Comment (KungFuDonkey) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp:334]: Please use the example of creating pairwise tests

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp [hunk 3]
```diff
@@ -4,26 +4,72 @@
...
+                                            << std::to_string(car.pub.DynGCg.pos.ay) << "\n" \
+                                            << std::to_string(car.pub.DynGCg.pos.az) << "\n" \
+                                            << std::to_string(car.pub.DynGC.vel.x) << "\n"   \
+                                            << std::to_string(car.pub.DynGC.acc.x) << "\n"   \
+                                            << std::to_string(car.priv.gear) << "\n";
+
+#define WRITE_EXPECTED_CONTROLS expected << std::to_string(car.ctrl.steer) << "\n"    \
+                                         << std::to_string(car.ctrl.brakeCmd) << "\n" \
+                                         << std::to_string(car.ctrl.accelCmd) << "\n" \
+                                         << std::to_string(car.ctrl.clutchCmd) << "\n"
```
Comment (KungFuDonkey) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp:71]: Change all '\n' characters with std::endl as enters are different on windows and linux

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp [hunk 4]
```diff
@@ -35,31 +81,327 @@ inline std::string GetTimeAsString(time_t p_date)
...
     FileDataStorage fileDataStorage;
-    fileDataStorage.Initialize(params, DUMMY_INITIALISATION_PARAMETERS);
-    fileDataStorage.Save(nullptr, nullptr, 0);
+    std::experimental::filesystem::path actualPath = fileDataStorage.Initialize(params, DUMMY_INITIALISATION_PARAMETERS);
+    expected << DUMMY_INITIALISATION_FILE_ENTRIES;
+
+    for (int i = 0; i < p_numberOfTicks; i++)
+    {
+        fileDataStorage.Save(nullptr, nullptr, 0);
+        expected << "0\n";
```
Comment (KungFuDonkey) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp:164]: Change all '\n' characters with std::endl as enters are different on windows and linux

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp [hunk 5]
```diff
@@ -35,31 +81,327 @@ inline std::string GetTimeAsString(time_t p_date)
...
+    GET_DUMMY_TIMES;
+    std::experimental::filesystem::path path = fileDataStorage.Initialize(params, DUMMY_INITIALISATION_PARAMETERS);
+    expected << DUMMY_INITIALISATION_FILE_ENTRIES;
+    if (p_storeEnvironment)
+    {
+        expected << "GameState\n";
+    }
+    if (p_storeControls)
+    {
+        expected << "UserInput\n";
```
Comment (KungFuDonkey) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp:217]: Change all '\n' characters with std::endl as enters are different on windows and linux

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp [hunk 6]
```diff
@@ -35,31 +81,327 @@ inline std::string GetTimeAsString(time_t p_date)
...
+    for (int i = 0; i < 3; i++)
+    {
+        // Save with random data
+        GET_RANDOM_TICKCOUNT;
+        GET_RANDOM_CAR;
+        GET_RANDOM_SITUATION;
+        fileDataStorage.Save(&car, &situation, tickCount);
+
+        // Define our expectations
+        expected << std::to_string(tickCount) << "\n";
```
Comment (KungFuDonkey) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp:229]: Change all '\n' characters with std::endl as enters are different on windows and linux

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp [hunk 7]
```diff
@@ -35,31 +81,327 @@ inline std::string GetTimeAsString(time_t p_date)
...
+    GET_DUMMY_TIMES;
+    std::experimental::filesystem::path path = fileDataStorage.Initialize(params, DUMMY_INITIALISATION_PARAMETERS);
+    expected << DUMMY_INITIALISATION_FILE_ENTRIES;
+
+    DecisionTuple decisions = GenerateDecisions(random, p_doSteer, p_doBrake, p_doAccel, p_doGear, p_doLights);
+
+    // Save any tick
+    GET_RANDOM_TICKCOUNT;
+    fileDataStorage.Save(nullptr, nullptr, tickCount);
+    expected << std::to_string(tickCount) << "\n";
```
Comment (KungFuDonkey) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp:284]: Change all '\n' characters with std::endl as enters are different on windows and linux

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp [hunk 8]
```diff
@@ -35,31 +81,327 @@ inline std::string GetTimeAsString(time_t p_date)
...
+    {
+        p_expected << "GearDecision\n"
+                   << std::to_string(p_decisions.GetGear()) << "\n";
+    }
+    if (p_doLights)
+    {
+        p_expected << "LightsDecision\n"
+                   << std::to_string(p_decisions.GetLights()) << "\n";
+    }
+    p_expected << "NONE\n";
```
Comment (KungFuDonkey) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp:140]: Change all '\n' characters with std::endl as enters are different on windows and linux

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp [hunk 9]
```diff
@@ -35,31 +81,327 @@ inline std::string GetTimeAsString(time_t p_date)
...
+    GET_DUMMY_TIMES;
+    std::experimental::filesystem::path path = fileDataStorage.Initialize(params, DUMMY_INITIALISATION_PARAMETERS);
+    expected << DUMMY_INITIALISATION_FILE_ENTRIES << "GameState\nUserInput\n";
+
+    // Generate and write data
+    GET_RANDOM_TICKCOUNT;
+    GET_RANDOM_CAR;
+    GET_RANDOM_SITUATION;
+    fileDataStorage.Save(&car, &situation, tickCount);
+    expected << std::to_string(tickCount) << "\n";
```
Comment (KungFuDonkey) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp:361]: Change all '\n' characters with std::endl as enters are different on windows and linux

**Review by KungFuDonkey — CHANGES_REQUESTED** [review]

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp (hunk 2 above)
Comment (Sindarin27) [reply speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backendUnitTests/FileDataStorageTests.cpp:331]: Example has already been followed.

**Review by KungFuDonkey — APPROVED** [review]

**Review by noutheesink — APPROVED** [review]

**Comment by KungFuDonkey** [comment]
@Sindarin27  is on vacation so I have merged this

**Pull request merged**